doc/pbh/pbh-design.md

```diff
@@ -0,0 +1,758 @@
+              / "INNER_L4_SRC_PORT"
+              / "INNER_DST_IPV4"
+              / "INNER_SRC_IPV4"
+              / "INNER_DST_IPV6"
+              / "INNER_SRC_IPV6"
+h16         = 1*4HEXDIG
+ls32        = h16 ":" h16
+dec-octet   = DIGIT                  ; 0-9
+              / %x31-39 DIGIT        ; 10-99
+              / %x31 2DIGIT          ; 100-199
+              / %x32 %x30-35 %x30-35 ; 200-255
+ipv4-prefix = dec-octet "." dec-octet "."
+              dec-octet "." dec-octet
+ipv6-prefix = 6( h16 ":" ) ls32
+              /                       "::" 5( h16 ":" ) ls32
+              / [               h16 ] "::" 4( h16 ":" ) ls32
+              / [ *1( h16 ":" ) h16 ] "::" 3( h16 ":" ) ls32
+              / [ *2( h16 ":" ) h16 ] "::" 2( h16 ":" ) ls32
+              / [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
+              / [ *4( h16 ":" ) h16 ] "::"              ls32
+              / [ *5( h16 ":" ) h16 ] "::"              h16
+              / [ *6( h16 ":" ) h16 ] "::"
+```
+
+**Note:** field _ipv4_mask_/_ipv6_mask_ is only valid when _hash_field_ equals _INNER_DST/SRC_IPV4/IPV6_
+
+### 2.4.2 Configuration sample
+
+**Inner 5-tuple hashing:**
+```json
+{
+    "PBH_HASH": {
+        "inner_ip_proto": {
+            "hash_field": "INNER_IP_PROTOCOL",
+            "sequence_id": "1"
+        },
+        "inner_l4_dst_port": {
+            "hash_field": "INNER_L4_DST_PORT",
+            "sequence_id": "2"
+        },
+        "inner_l4_src_port": {
+            "hash_field": "INNER_L4_SRC_PORT",
+            "sequence_id": "2"
+        },
+        "inner_dst_ipv4": {
+            "hash_field": "INNER_DST_IPV4",
+            "ipv4_mask": "255.0.0.0",
+            "sequence_id": "3"
+        },
+        "inner_src_ipv4": {
+            "hash_field": "INNER_SRC_IPV4",
+            "ipv4_mask": "0.0.0.255",
+            "sequence_id": "3"
+        },
+        "inner_dst_ipv6": {
+            "hash_field": "INNER_DST_IPV6",
+            "ipv6_mask": "FFFF::",
+            "sequence_id": "4"
+        },
+        "inner_src_ipv6": {
+            "hash_field": "INNER_SRC_IPV6",
+            "ipv6_mask": "::FFFF",
+            "sequence_id": "4"
+        }
+    },
+    "PBH_RULE": {
+        "pbh_table|nvgre": {
+            "priority": "1",
+            "gre_key": "0x2500/0xffffff00",
+            "inner_ether_type": "0x86dd/0xffff",
+            "hash_list": [
+                "inner_ip_proto",
+                "inner_l4_dst_port",
+                "inner_l4_src_port",
+                "inner_dst_ipv6",
+                "inner_src_ipv6"
+            ],
+            "packet_action": "SET_ECMP_HASH"
+        },
+        "pbh_table|vxlan": {
+            "priority": "2",
+            "ip_protocol": "0x11/0xff",
+            "l4_dst_port": "0x12b5/0xffff",
+            "inner_ether_type": "0x0800/0xffff",
+            "hash_list": [
+                "inner_ip_proto",
+                "inner_l4_dst_port",
+                "inner_l4_src_port",
```

Comment:
If hash_list is a leaf-list, the values should be ["PBH_HASH|inner_ip_proto", "PBH_HASH|inner_l4_dst_port"...], rather we can have the following, please see if this makes sense, if not, please follow appropriate YANG conventions to represent the references.
1) PBH_TABLE|vxlan 
     hash_list ---leaf-ref---> PBH_HASH_SET|5tuple

2) PBH_HASH_SET|5tuple

3)
PBH_HASH|5tuple|inner_ip_proto
PBH_HASH|5tuple|inner_l4_dst_port
...
....

Reply:
@venkatmahalingam according to the SWSS DB schema (https://github.com/Azure/sonic-swss/blob/master/doc/swss-schema.md) we have a concept of key reference. Today in SONiC it is being used in few places (qos, buffer, pg, queue, etc.):

**Example:**
```
    "PORT_QOS_MAP": {
        "Ethernet8": {
            "tc_to_queue_map": "[TC_TO_QUEUE_MAP|AZURE]",
            "dscp_to_tc_map": "[DSCP_TO_TC_MAP|AZURE]",
            "pfc_enable": "3,4",
            "pfc_to_queue_map": "[MAP_PFC_PRIORITY_TO_QUEUE|AZURE]",
            "pfc_to_pg_map": "[PFC_PRIORITY_TO_PRIORITY_GROUP_MAP|AZURE]",
            "tc_to_pg_map": "[TC_TO_PRIORITY_GROUP_MAP|AZURE]"
        },

    "BUFFER_PORT_INGRESS_PROFILE_LIST": {
        "Ethernet8": {
            "profile_list": "[BUFFER_PROFILE|ingress_lossless_profile]"
        },

    "BUFFER_PORT_EGRESS_PROFILE_LIST": {
        "Ethernet8": {
            "profile_list": "[BUFFER_PROFILE|egress_lossless_profile],[BUFFER_PROFILE|egress_lossy_profile]"
        },

    "BUFFER_PG": {
        "Ethernet14|0": {
            "profile": "[BUFFER_PROFILE|ingress_lossy_profile]"
        },

    "BUFFER_QUEUE": {
        "Ethernet12|5-6": {
            "profile": "[BUFFER_PROFILE|q_lossy_profile]"
        },

    "QUEUE": {
        "Ethernet20|5": {
            "scheduler": "[SCHEDULER|scheduler.0]"
        },
        "Ethernet48|5": {
            "scheduler": "[SCHEDULER|scheduler.0]"
        },
        "Ethernet48|4": {
            "wred_profile": "[WRED_PROFILE|AZURE_LOSSLESS]",
            "scheduler": "[SCHEDULER|scheduler.1]"
        },
```

With a new YANG validation concept, it seem to be redundant to use this style, as it makes sense only for ABNF notation.
Please have a look at this YANG model: https://github.com/Azure/sonic-buildimage/pull/7461.

It was written following the existing ACL `ports` leaf-list concept:
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-acl.yang#L257
```yang
leaf-list ports {
	/* union of leafref is allowed in YANG 1.1 */
	type union {
		type leafref {
			path /port:sonic-port/port:PORT/port:PORT_LIST/port:name;
		}
		type leafref {
			path /lag:sonic-portchannel/lag:PORTCHANNEL/lag:PORTCHANNEL_LIST/lag:name;
		}
		type string {
			pattern "";
		}
	}
	/* Today in SONiC, we do not delete the list once
	 * created, instead we set to empty list. Due to that
	 * below default values are needed.
	 */
	default "";
}
```

And this is how the ACL/PORT/LAG configuration looks like in `config_db.json`:
```json
"ACL_TABLE": {
    "EVERFLOW": {
        "ports": [
            "PortChannel0001",
            "PortChannel0002",
            "PortChannel0003",
            "PortChannel0004",
            "Ethernet110",
            "Ethernet108",
            "Ethernet24",
            "Ethernet26",
            "Ethernet12",
            "Ethernet14",
            "Ethernet62",
            "Ethernet60",
            "Ethernet46",
            "Ethernet44",
            "Ethernet78",
            "Ethernet76",
            "Ethernet48",
            "Ethernet28",
            "Ethernet116",
            "Ethernet74",
            "Ethernet72",
            "Ethernet30",
            "Ethernet80",
            "Ethernet82",
            "Ethernet120",
            "Ethernet18",
            "Ethernet16",
            "Ethernet122",
            "Ethernet58",
            "Ethernet56",
            "Ethernet50",
            "Ethernet96",
            "Ethernet112",
            "Ethernet52",
            "Ethernet54",
            "Ethernet84",
            "Ethernet124",
            "Ethernet22",
            "Ethernet20",
            "Ethernet64"
        ],
        "type": "MIRROR",
        "policy_desc": "EVERFLOW",
        "stage": "ingress"
    }
}

"PORT": {
    "Ethernet80": {
        "index": "21",
        "lanes": "80,81",
        "description": "Servers61:eth0",
        "pfc_asym": "off",
        "mtu": "9100",
        "alias": "etp21a",
        "admin_status": "up",
        "speed": "50000"
    },
    "Ethernet224": {
        "index": "57",
        "lanes": "224,225",
        "description": "etp57a",
        "mtu": "9100",
        "alias": "etp57a",
        "pfc_asym": "off",
        "speed": "50000"
    }
}

"PORTCHANNEL": {
    "PortChannel0001": {
        "admin_status": "up",
        "min_links": "2",
        "members": [
            "Ethernet0",
            "Ethernet2"
        ],
        "mtu": "9100"
    },
    "PortChannel0003": {
        "admin_status": "up",
        "min_links": "2",
        "members": [
            "Ethernet32",
            "Ethernet34"
        ],
        "mtu": "9100"
    }
}
```

As you may see, it doesn't use `["PORT|Ethernet80", "PORTCHANNEL|PortChannel0001", ...]` notation.

Note: PBH YANG model was tested with DPB feature - no issues observed

Reply:
ok, thanks.